### ISSUE TITLE
Simplify dependencies for basic gridding

### ISSUE THREAD

**Issue by djhoese** [deeplycloudy/glmtools#49]
I've been reviewing some of the glmtools code to see what would be needed to bundle it in a CSPP (Community Satellite Processing Package) glmtools package. I had a couple things I was wondering about or changes I'd like to make but wanted to run them by glmtools developer(s). So far this is all about dependencies. We have experience slimming down and compiling some common python dependencies but not others and others are just kind of large so if they aren't absolutely needed it would be nice to save the size in the final bundle that is released.

1. `matplotlib` is used for what seems like polygon calculations. Is there a smaller library that could perform these operations? Can `shapely` do this? It depends on the GEOS lib, so I'm not sure that's a net gain on compilation complexity.
2. `tables` (pytables package) is only used in certain functions and isn't actually needed by glmtools directly from what I can tell. The biggest complication was lmatools using it in a class for reading HDF5 files. glmtools subclasses from this class and thus imports tables but never actually uses HDF5. I was able to successfully move the tables imports to various functions and got it to work; at least I think I did.
3. `scikit-learn` seems to be used only for it's KDTree implementation. I wonder if a lighter weight library like `pykdtree` could we used instead? I will try running some tests later today if I can and see how it compares (if the results are the same). We use pykdtree in pyresample.
4. The `environment.yaml` includes `ipython` as a dependency which is not strictly a dependence of `glmtools`. This is just an FYI, but not a request for any change.
5. Pyclipper is ~not~now available on conda-forge so no need to manually pip install it (not sure if you had a part in this or not).

**Comment by dopplershift** [comment]
> Pyclipper is not available on conda-forge so no need to manually pip install it

Huh?

**Comment by djhoese** [comment]
*now* not *not*

**Comment by deeplycloudy** [comment]
1. I think all of the polygon math in glmtools is handled by pyclipper, and glmtools only uses matplotlib for visualization. There is a matplotlib dependency in lmatools, but [I've been meaning to fence that off for some time](https://github.com/deeplycloudy/lmatools/issues/23).

2. Yes, `tables` is not used from glmtools. If we can move things around in lmatools to fence it, I'm good with that.

3. even better … it is only used in one function that from an earlier development effort and is no longer used. The import has been fenced off in [d328924](https://github.com/deeplycloudy/glmtools/commit/d328924eec5652d41ed6812986545af87576ede2) and so `scikit-learn` could be removed from `environment.yml`. 

4. Yes, it was there for my convenience. If this becomes a problem down the road for packaging we can remove it.

5. Oh good, I'm glad to hear pyclipper is now on conda-forge. An update to the `environment.yml` and docs would be welcome.

More broadly, there is a lot of orphaned code and untested paths of operation (e.g., gridding directly to a lat lon grid) that hasn't been tested in a long time, and is probably by now quite slow. Some pruning could reduce the actual LOC in glmtools, perhaps quite substantially. Likewise, I've wanted to explore refactoring or even vendoring parts of lmatools to eliminate the need to install it.

**Comment by djhoese** [comment]
Wow, all great stuff to hear. None of this is necessary to bundle glmtools in a CSPP package/tarball, but like I said if we've packaged all of them before then it is less work to make sure they compile well. I'm sure some users would like a smaller tarball too.

So if I start from the ugf-grid branch, I'll need to merge master in for the scikit fencing, right? I'll look in to the matplotlib import/dependency and make sure it isn't needed. If it is imported I can make a pull request where needed.

I'll see what I can get working. Thanks.

**Comment by deeplycloudy** [comment]
The ugf-newgrid branch should have that fencing from master. Both it and the isatss branch are kept sync’d through master as parallel “production” branches.

**Comment by djhoese** [comment]
@deeplycloudy Does this sound right that matplotlib needs to be fenced off in stormdrain?

```
    from glmtools.grid.make_grids import grid_GLM_flashes
  File "/opt/ssec/cspp_geo_gridded_glm-v1.0.0-prealpha0/opt/conda/lib/python3.7/site-packages/glmtools/grid/make_grids.py", line 13, in <module>
    from glmtools.io.mimic_lma import read_flashes
  File "/opt/ssec/cspp_geo_gridded_glm-v1.0.0-prealpha0/opt/conda/lib/python3.7/site-packages/glmtools/io/mimic_lma.py", line 19, in <module>
    from lmatools.lasso.energy_stats import TimeSeriesPolygonLassoFilter
  File "/opt/ssec/cspp_geo_gridded_glm-v1.0.0-prealpha0/opt/conda/lib/python3.7/site-packages/lmatools/lasso/energy_stats.py", line 18, in <module>
    from stormdrain.support.matplotlib.formatters import SecDayFormatter
  File "/opt/ssec/cspp_geo_gridded_glm-v1.0.0-prealpha0/opt/conda/lib/python3.7/site-packages/stormdrain/support/matplotlib/formatters.py", line 5, in <module>
    from matplotlib.ticker import Formatter, EngFormatter
ModuleNotFoundError: No module named 'matplotlib'
```

Could you also explain what the different branches are for? Are they meant to stay separate or will they be merged to master in the future?

**Comment by djhoese** [comment]
With #56 this seems to be all figured out. Any plans to merge `minvaluegrids` to lmatools master? I suppose it would require `ugf-newgrid` on glmtools to be merged with master?

**Comment by deeplycloudy** [comment]
@djhoese I'm waiting to merge those branches to master until the NWS-facing use of glmtools is ready to move to `ugf-newgrid`. I think that may happen in the next couple weeks. Having merged #56, is there any reason to keep this issue open?
